{"nbformat":4,"nbformat_minor":0,"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.12"},"colab":{"name":"plant_gene_expression_kmapper.ipynb","provenance":[]}},"cells":[{"cell_type":"markdown","metadata":{"id":"jFDzb40UH1XU"},"source":["## Applying mapper to plant gene expression data\n","\n","In this notebook, we will apply mapper algorithm to the gene expression data collected by last year's class.<br>\n","Download all the files from the shared google drive folder into a directory on your computer if you are running jupyter notebooks locally. The data is stored in two csv files, one contains the gene expression profiles, the other contains the metadata such as sample id, family, tissue type, stress, etc.<br> Along with the data files, there are two notebooks (including this one!) and two python script files which we will need. Make sure you have all these files downloaded into the same directory."]},{"cell_type":"markdown","metadata":{"id":"-rbnML1QH1XY"},"source":["### Imort useful packages / modules"]},{"cell_type":"code","metadata":{"id":"jfcdSiLXH1XY","executionInfo":{"status":"ok","timestamp":1636479653391,"user_tz":360,"elapsed":746,"user":{"displayName":"Damián Villaseñor Amador","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Ggqk0xsm9537Ect6EKfhPX7AEk63wwoSTLmiDrW5A=s64","userId":"18379377199811291117"}}},"source":["import sys\n","import numpy as np\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","\n","# ML tools\n","from sklearn.cluster import DBSCAN\n","from sklearn.preprocessing import MinMaxScaler\n","\n","# For output display\n","from IPython.display import IFrame\n","\n","# If running locally, set current directory as projdir\n","projdir = '.'"],"execution_count":1,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"97hBlse7OqZe"},"source":["### Mount Google Drive*\n","\n","Run the next cell __only if you're planning to run this notebook on Google Colab__. If you are running this notebook locally, comment it out.\n","\n","To run the notebook in Google Colab, it is best if you upload data and script files to a folder in Google Drive and access them from there. We already have a shared Google Drive folder named `PlantsAndPython-2021-10-22` that contains all the required data and script files.\n","Run the next code cell to mount the drive and make the files accessible. We will define the shared folder as our project directory `projdir`.\n","\n","If Google drive is not already mounted, running the cell will produce a link. Click on the link, follow the prompts to log in to your google account, and copy the text string generated at the end. Paste the text string in the box below and press `Enter`."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"WEpUU6zvIDpQ","executionInfo":{"status":"ok","timestamp":1636479702327,"user_tz":360,"elapsed":30537,"user":{"displayName":"Damián Villaseñor Amador","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Ggqk0xsm9537Ect6EKfhPX7AEk63wwoSTLmiDrW5A=s64","userId":"18379377199811291117"}},"outputId":"03920fcc-f71d-459f-b5c1-f46b43826c39"},"source":["# Only if running in Google Colab..!!\n","# DO NOT run this cell if running locally - simply comment it out.\n","from google.colab import drive\n","drive.mount('/content/gdrive')\n","\n","projdir = '/content/gdrive/MyDrive/PlantsAndPython-2021-10-22'\n","sys.path.append(projdir)"],"execution_count":3,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/gdrive\n"]}]},{"cell_type":"code","metadata":{"id":"bfVRuZwPIkAo","colab":{"base_uri":"https://localhost:8080/","height":374},"executionInfo":{"status":"error","timestamp":1636479704984,"user_tz":360,"elapsed":349,"user":{"displayName":"Damián Villaseñor Amador","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Ggqk0xsm9537Ect6EKfhPX7AEk63wwoSTLmiDrW5A=s64","userId":"18379377199811291117"}},"outputId":"74013885-e94b-4c01-c5cb-5fd012ba14fe"},"source":["# import helper_functions\n","from helper_functions import loaddata\n","from helper_functions import colorscale_from_matplotlib_cmap\n","\n","# import lense function\n","from lenses import fsga_transform\n","\n","# keppler mapper\n","import kmapper as km"],"execution_count":4,"outputs":[{"output_type":"error","ename":"ModuleNotFoundError","evalue":"ignored","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)","\u001b[0;32m<ipython-input-4-458d4e1c215d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# import helper_functions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mhelper_functions\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mloaddata\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mhelper_functions\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mcolorscale_from_matplotlib_cmap\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# import lense function\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'helper_functions'","","\u001b[0;31m---------------------------------------------------------------------------\u001b[0;32m\nNOTE: If your import is failing due to a missing package, you can\nmanually install dependencies using either !pip or !apt.\n\nTo view examples of installing some common dependencies, click the\n\"Open Examples\" button below.\n\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n"],"errorDetails":{"actions":[{"action":"open_url","actionText":"Open Examples","url":"/notebooks/snippets/importing_libraries.ipynb"}]}}]},{"cell_type":"markdown","metadata":{"id":"xhnDinWmH1Xa"},"source":["### Data\n","\n","The data is stored in two csv files. `clean_metadata.csv` contains the metadata, `clean_RNAseq_sv_corrected.csv` contains the gene expression data. We are interested in three factors in particular: family, tissue type and stress type."]},{"cell_type":"code","metadata":{"id":"--t6eU5tH1Xa"},"source":["factorfile = projdir + '/clean_metadata.csv'\n","rnafile = projdir + '/clean_RNAseq_sv_corrected.csv'\n","\n","factors = ['stress', 'tissue', 'family']\n","levels = ['healthy', 'leaf', 'Poaceae']\n","\n","filter_by_factor, filter_by_level = ('family', 'Poaceae')\n","color_by_factor, color_by_level = ('tissue', 'leaf')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"J8nZD0i_H1Xb"},"source":["We will use the custom `loaddata` function to load the data from the csv files and merge them using the *sra*s. Feel free to take a look at the function defined in file `helper_functions.py`. The number of SRAs in the two files is different. We will only take SRAs for which we have both the gene expression profile and the factor data available. This is done using the `merge` method from pandas."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"B2mUzLSnH1Xb","executionInfo":{"status":"ok","timestamp":1636389898526,"user_tz":300,"elapsed":9017,"user":{"displayName":"Sourabh Palande","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"18429378630788159865"}},"outputId":"76eddaef-c68d-429d-a013-d7481b1ce5b7"},"source":["df, orthos = loaddata(factorfile, rnafile, factors)"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Factor input data shape: (3172, 7)\n","RNA input data shape: (6335, 2672)\n","Dataframe shape after merge: (2671, 6342)\n"]}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":252},"id":"yVIeEosWH1Xc","executionInfo":{"status":"ok","timestamp":1636389898655,"user_tz":300,"elapsed":18,"user":{"displayName":"Sourabh Palande","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"18429378630788159865"}},"outputId":"ceae4998-c828-408f-bb29-590fd2f86e4d"},"source":["df.head()"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>sra</th>\n","      <th>sample_id</th>\n","      <th>species</th>\n","      <th>family</th>\n","      <th>tissue</th>\n","      <th>stress</th>\n","      <th>bioproject</th>\n","      <th>OG0002118</th>\n","      <th>OG0002119</th>\n","      <th>OG0002120</th>\n","      <th>OG0002121</th>\n","      <th>OG0002122</th>\n","      <th>OG0002125</th>\n","      <th>OG0002126</th>\n","      <th>OG0002127</th>\n","      <th>OG0002128</th>\n","      <th>OG0002132</th>\n","      <th>OG0002133</th>\n","      <th>OG0002137</th>\n","      <th>OG0002139</th>\n","      <th>OG0002142</th>\n","      <th>OG0002143</th>\n","      <th>OG0002144</th>\n","      <th>OG0002145</th>\n","      <th>OG0002146</th>\n","      <th>OG0002147</th>\n","      <th>OG0002148</th>\n","      <th>OG0002154</th>\n","      <th>OG0002155</th>\n","      <th>OG0002156</th>\n","      <th>OG0002157</th>\n","      <th>OG0002159</th>\n","      <th>OG0002160</th>\n","      <th>OG0002161</th>\n","      <th>OG0002162</th>\n","      <th>OG0002165</th>\n","      <th>OG0002166</th>\n","      <th>OG0002167</th>\n","      <th>OG0002169</th>\n","      <th>OG0002170</th>\n","      <th>...</th>\n","      <th>OG0011191</th>\n","      <th>OG0011201</th>\n","      <th>OG0011202</th>\n","      <th>OG0011207</th>\n","      <th>OG0011212</th>\n","      <th>OG0011216</th>\n","      <th>OG0011238</th>\n","      <th>OG0011240</th>\n","      <th>OG0011261</th>\n","      <th>OG0011264</th>\n","      <th>OG0011266</th>\n","      <th>OG0011267</th>\n","      <th>OG0011271</th>\n","      <th>OG0011272</th>\n","      <th>OG0011273</th>\n","      <th>OG0011275</th>\n","      <th>OG0011277</th>\n","      <th>OG0011279</th>\n","      <th>OG0011282</th>\n","      <th>OG0011284</th>\n","      <th>OG0011285</th>\n","      <th>OG0011328</th>\n","      <th>OG0011329</th>\n","      <th>OG0011330</th>\n","      <th>OG0011335</th>\n","      <th>OG0011346</th>\n","      <th>OG0011360</th>\n","      <th>OG0011364</th>\n","      <th>OG0011366</th>\n","      <th>OG0011368</th>\n","      <th>OG0011413</th>\n","      <th>OG0011424</th>\n","      <th>OG0011438</th>\n","      <th>OG0011439</th>\n","      <th>OG0011503</th>\n","      <th>OG0011580</th>\n","      <th>OG0011583</th>\n","      <th>OG0011599</th>\n","      <th>OG0011603</th>\n","      <th>OG0011655</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>SRR1598911</td>\n","      <td>A_hypochondria_001</td>\n","      <td>A_hypochondria</td>\n","      <td>Amaranthacea</td>\n","      <td>flower</td>\n","      <td>healthy</td>\n","      <td>PRJNA263128</td>\n","      <td>-0.001456</td>\n","      <td>0.105937</td>\n","      <td>-0.050513</td>\n","      <td>-0.064531</td>\n","      <td>0.041956</td>\n","      <td>0.001421</td>\n","      <td>0.015684</td>\n","      <td>-0.044681</td>\n","      <td>-0.040176</td>\n","      <td>0.156901</td>\n","      <td>0.009040</td>\n","      <td>0.086616</td>\n","      <td>0.198255</td>\n","      <td>-0.011510</td>\n","      <td>0.116082</td>\n","      <td>0.153478</td>\n","      <td>-0.048863</td>\n","      <td>0.031323</td>\n","      <td>-0.062965</td>\n","      <td>-0.027801</td>\n","      <td>0.059899</td>\n","      <td>-0.018740</td>\n","      <td>-0.027129</td>\n","      <td>-0.070887</td>\n","      <td>-0.081776</td>\n","      <td>-0.035525</td>\n","      <td>-0.072955</td>\n","      <td>-0.017868</td>\n","      <td>-0.048536</td>\n","      <td>-0.048135</td>\n","      <td>-0.106876</td>\n","      <td>-0.072206</td>\n","      <td>-0.101233</td>\n","      <td>...</td>\n","      <td>-0.039385</td>\n","      <td>-0.018324</td>\n","      <td>-0.072203</td>\n","      <td>0.018669</td>\n","      <td>-0.079136</td>\n","      <td>0.006234</td>\n","      <td>-0.055142</td>\n","      <td>-0.074703</td>\n","      <td>-0.051067</td>\n","      <td>-0.003863</td>\n","      <td>-0.073836</td>\n","      <td>-0.009686</td>\n","      <td>-0.071781</td>\n","      <td>-0.051653</td>\n","      <td>-0.082667</td>\n","      <td>-0.055877</td>\n","      <td>-0.070845</td>\n","      <td>-0.107171</td>\n","      <td>-0.062462</td>\n","      <td>-0.068771</td>\n","      <td>-0.070214</td>\n","      <td>-0.064071</td>\n","      <td>-0.067258</td>\n","      <td>-0.055369</td>\n","      <td>-0.073002</td>\n","      <td>-0.075120</td>\n","      <td>-0.067491</td>\n","      <td>-0.074643</td>\n","      <td>-0.055568</td>\n","      <td>0.068693</td>\n","      <td>-0.044161</td>\n","      <td>-0.065706</td>\n","      <td>-0.073235</td>\n","      <td>-0.043796</td>\n","      <td>-0.072423</td>\n","      <td>-0.051586</td>\n","      <td>0.034663</td>\n","      <td>-0.046609</td>\n","      <td>0.041908</td>\n","      <td>0.020650</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>SRR1598912</td>\n","      <td>A_hypochondria_002</td>\n","      <td>A_hypochondria</td>\n","      <td>Amaranthacea</td>\n","      <td>leaf</td>\n","      <td>healthy</td>\n","      <td>PRJNA263128</td>\n","      <td>-0.050352</td>\n","      <td>0.254496</td>\n","      <td>-0.229081</td>\n","      <td>-0.044667</td>\n","      <td>-0.029998</td>\n","      <td>0.093129</td>\n","      <td>-0.050958</td>\n","      <td>-0.237801</td>\n","      <td>-0.166522</td>\n","      <td>0.274920</td>\n","      <td>-0.092045</td>\n","      <td>0.313335</td>\n","      <td>0.687212</td>\n","      <td>0.170770</td>\n","      <td>0.767898</td>\n","      <td>0.421275</td>\n","      <td>-0.202901</td>\n","      <td>0.198959</td>\n","      <td>-0.188598</td>\n","      <td>-0.242519</td>\n","      <td>0.426952</td>\n","      <td>-0.134623</td>\n","      <td>-0.151437</td>\n","      <td>-0.183163</td>\n","      <td>-0.133909</td>\n","      <td>-0.063276</td>\n","      <td>-0.220159</td>\n","      <td>-0.088992</td>\n","      <td>-0.156517</td>\n","      <td>-0.134450</td>\n","      <td>-0.052810</td>\n","      <td>-0.203709</td>\n","      <td>0.082189</td>\n","      <td>...</td>\n","      <td>-0.245452</td>\n","      <td>0.509246</td>\n","      <td>-0.226191</td>\n","      <td>0.078928</td>\n","      <td>-0.268059</td>\n","      <td>-0.424409</td>\n","      <td>-0.235317</td>\n","      <td>-0.241414</td>\n","      <td>-0.201529</td>\n","      <td>0.001640</td>\n","      <td>-0.248217</td>\n","      <td>0.163692</td>\n","      <td>-0.242656</td>\n","      <td>-0.227221</td>\n","      <td>-0.252562</td>\n","      <td>-0.233738</td>\n","      <td>-0.234675</td>\n","      <td>-0.366289</td>\n","      <td>-0.237922</td>\n","      <td>-0.235524</td>\n","      <td>-0.234050</td>\n","      <td>-0.223131</td>\n","      <td>-0.258069</td>\n","      <td>-0.140590</td>\n","      <td>-0.230676</td>\n","      <td>-0.254929</td>\n","      <td>-0.221565</td>\n","      <td>-0.255772</td>\n","      <td>-0.157762</td>\n","      <td>0.635246</td>\n","      <td>-0.247876</td>\n","      <td>-0.235771</td>\n","      <td>-0.224390</td>\n","      <td>-0.271840</td>\n","      <td>-0.217694</td>\n","      <td>0.217500</td>\n","      <td>-0.084585</td>\n","      <td>-0.123412</td>\n","      <td>0.621273</td>\n","      <td>-0.098835</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>SRR1598913</td>\n","      <td>A_hypochondria_003</td>\n","      <td>A_hypochondria</td>\n","      <td>Amaranthacea</td>\n","      <td>root</td>\n","      <td>healthy</td>\n","      <td>PRJNA263128</td>\n","      <td>-0.207445</td>\n","      <td>0.485422</td>\n","      <td>-0.300600</td>\n","      <td>-0.085825</td>\n","      <td>-0.044407</td>\n","      <td>-0.061632</td>\n","      <td>-0.047539</td>\n","      <td>-0.299363</td>\n","      <td>-0.107795</td>\n","      <td>0.566231</td>\n","      <td>-0.330444</td>\n","      <td>0.275767</td>\n","      <td>0.763075</td>\n","      <td>0.401680</td>\n","      <td>0.434029</td>\n","      <td>0.450636</td>\n","      <td>-0.278571</td>\n","      <td>0.352884</td>\n","      <td>-0.218325</td>\n","      <td>-0.206967</td>\n","      <td>0.202653</td>\n","      <td>-0.136491</td>\n","      <td>-0.082368</td>\n","      <td>-0.406822</td>\n","      <td>0.224917</td>\n","      <td>-0.139774</td>\n","      <td>-0.311683</td>\n","      <td>-0.132602</td>\n","      <td>-0.203021</td>\n","      <td>-0.334327</td>\n","      <td>-0.271979</td>\n","      <td>-0.256626</td>\n","      <td>0.113280</td>\n","      <td>...</td>\n","      <td>-0.179593</td>\n","      <td>-0.117367</td>\n","      <td>-0.286509</td>\n","      <td>-0.031956</td>\n","      <td>-0.329782</td>\n","      <td>0.462999</td>\n","      <td>-0.202286</td>\n","      <td>-0.339019</td>\n","      <td>-0.251814</td>\n","      <td>0.067319</td>\n","      <td>-0.306012</td>\n","      <td>0.073821</td>\n","      <td>-0.309367</td>\n","      <td>-0.187119</td>\n","      <td>-0.203944</td>\n","      <td>-0.296097</td>\n","      <td>-0.311260</td>\n","      <td>-0.046690</td>\n","      <td>-0.310558</td>\n","      <td>-0.279481</td>\n","      <td>-0.303574</td>\n","      <td>-0.265100</td>\n","      <td>-0.297655</td>\n","      <td>-0.227918</td>\n","      <td>-0.307558</td>\n","      <td>-0.313380</td>\n","      <td>-0.296619</td>\n","      <td>-0.314493</td>\n","      <td>-0.215667</td>\n","      <td>0.049767</td>\n","      <td>-0.187849</td>\n","      <td>-0.303212</td>\n","      <td>-0.283030</td>\n","      <td>0.087385</td>\n","      <td>-0.288310</td>\n","      <td>-0.013373</td>\n","      <td>0.024199</td>\n","      <td>-0.195171</td>\n","      <td>0.360540</td>\n","      <td>0.137576</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>SRR1598910</td>\n","      <td>A_hypochondria_004</td>\n","      <td>A_hypochondria</td>\n","      <td>Amaranthacea</td>\n","      <td>stem</td>\n","      <td>healthy</td>\n","      <td>PRJNA263128</td>\n","      <td>-0.226247</td>\n","      <td>0.661615</td>\n","      <td>-0.122427</td>\n","      <td>-0.194361</td>\n","      <td>-0.018764</td>\n","      <td>0.062965</td>\n","      <td>-0.090751</td>\n","      <td>-0.343471</td>\n","      <td>-0.203417</td>\n","      <td>0.860485</td>\n","      <td>-0.163096</td>\n","      <td>0.596404</td>\n","      <td>1.241189</td>\n","      <td>0.574202</td>\n","      <td>0.576262</td>\n","      <td>1.291220</td>\n","      <td>-0.341597</td>\n","      <td>0.507157</td>\n","      <td>-0.234809</td>\n","      <td>-0.328655</td>\n","      <td>0.425913</td>\n","      <td>-0.143535</td>\n","      <td>-0.070677</td>\n","      <td>-0.411144</td>\n","      <td>0.061430</td>\n","      <td>-0.076929</td>\n","      <td>-0.348331</td>\n","      <td>-0.146977</td>\n","      <td>-0.183348</td>\n","      <td>-0.408786</td>\n","      <td>-0.013172</td>\n","      <td>-0.259845</td>\n","      <td>-0.747038</td>\n","      <td>...</td>\n","      <td>-0.218861</td>\n","      <td>-0.064852</td>\n","      <td>-0.326633</td>\n","      <td>-0.081825</td>\n","      <td>-0.371809</td>\n","      <td>0.261899</td>\n","      <td>-0.293581</td>\n","      <td>-0.339895</td>\n","      <td>-0.265222</td>\n","      <td>0.301502</td>\n","      <td>-0.350525</td>\n","      <td>-0.358029</td>\n","      <td>-0.345393</td>\n","      <td>-0.280741</td>\n","      <td>-0.241136</td>\n","      <td>-0.330304</td>\n","      <td>-0.350639</td>\n","      <td>-0.219577</td>\n","      <td>-0.346377</td>\n","      <td>-0.313366</td>\n","      <td>-0.347531</td>\n","      <td>-0.300973</td>\n","      <td>-0.344449</td>\n","      <td>-0.271111</td>\n","      <td>-0.356876</td>\n","      <td>-0.357422</td>\n","      <td>-0.344556</td>\n","      <td>-0.359829</td>\n","      <td>-0.258149</td>\n","      <td>0.297424</td>\n","      <td>-0.279176</td>\n","      <td>-0.335267</td>\n","      <td>-0.326472</td>\n","      <td>-0.125061</td>\n","      <td>-0.335647</td>\n","      <td>-0.067140</td>\n","      <td>-0.054410</td>\n","      <td>-0.166659</td>\n","      <td>-0.174771</td>\n","      <td>-0.063741</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>SRR1598914</td>\n","      <td>A_hypochondria_005</td>\n","      <td>A_hypochondria</td>\n","      <td>Amaranthacea</td>\n","      <td>other</td>\n","      <td>drought</td>\n","      <td>PRJNA263128</td>\n","      <td>-0.103369</td>\n","      <td>1.088749</td>\n","      <td>-0.300173</td>\n","      <td>-0.028792</td>\n","      <td>0.336153</td>\n","      <td>0.015617</td>\n","      <td>0.051135</td>\n","      <td>-0.303071</td>\n","      <td>-0.252303</td>\n","      <td>1.050839</td>\n","      <td>0.133557</td>\n","      <td>-0.013915</td>\n","      <td>1.288096</td>\n","      <td>0.023644</td>\n","      <td>0.630677</td>\n","      <td>0.569392</td>\n","      <td>-0.189553</td>\n","      <td>0.297558</td>\n","      <td>-0.254742</td>\n","      <td>-0.247522</td>\n","      <td>0.157631</td>\n","      <td>-0.158832</td>\n","      <td>-0.045626</td>\n","      <td>-0.323843</td>\n","      <td>-0.267552</td>\n","      <td>-0.122531</td>\n","      <td>-0.263435</td>\n","      <td>-0.009662</td>\n","      <td>-0.186143</td>\n","      <td>-0.185139</td>\n","      <td>-0.188903</td>\n","      <td>-0.262444</td>\n","      <td>0.365195</td>\n","      <td>...</td>\n","      <td>-0.256589</td>\n","      <td>0.127434</td>\n","      <td>-0.279744</td>\n","      <td>0.087709</td>\n","      <td>-0.343602</td>\n","      <td>0.066369</td>\n","      <td>-0.213633</td>\n","      <td>-0.324498</td>\n","      <td>-0.171402</td>\n","      <td>-0.405118</td>\n","      <td>-0.324843</td>\n","      <td>-0.338195</td>\n","      <td>-0.322803</td>\n","      <td>-0.201093</td>\n","      <td>-0.236512</td>\n","      <td>-0.249201</td>\n","      <td>-0.308489</td>\n","      <td>-0.573443</td>\n","      <td>-0.313946</td>\n","      <td>-0.284210</td>\n","      <td>-0.305934</td>\n","      <td>-0.276824</td>\n","      <td>-0.293528</td>\n","      <td>-0.243731</td>\n","      <td>-0.323496</td>\n","      <td>-0.325155</td>\n","      <td>-0.294073</td>\n","      <td>-0.322954</td>\n","      <td>-0.248349</td>\n","      <td>0.551140</td>\n","      <td>-0.183794</td>\n","      <td>-0.295979</td>\n","      <td>-0.302226</td>\n","      <td>-0.088812</td>\n","      <td>-0.278099</td>\n","      <td>0.031540</td>\n","      <td>0.025193</td>\n","      <td>-0.184997</td>\n","      <td>-0.522966</td>\n","      <td>0.173639</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>5 rows × 6342 columns</p>\n","</div>"],"text/plain":["          sra           sample_id  ... OG0011603 OG0011655\n","0  SRR1598911  A_hypochondria_001  ...  0.041908  0.020650\n","1  SRR1598912  A_hypochondria_002  ...  0.621273 -0.098835\n","2  SRR1598913  A_hypochondria_003  ...  0.360540  0.137576\n","3  SRR1598910  A_hypochondria_004  ... -0.174771 -0.063741\n","4  SRR1598914  A_hypochondria_005  ... -0.522966  0.173639\n","\n","[5 rows x 6342 columns]"]},"metadata":{},"execution_count":6}]},{"cell_type":"markdown","metadata":{"id":"vCMdb637H1Xd"},"source":["`df` is the dataframe containing the merged factor and RNASeq data. `orthos` is just a list of orthogroup names that will be useful when we want to select only the part of dataframe containing RNASeq data."]},{"cell_type":"markdown","metadata":{"id":"MZXonrk2H1Xd"},"source":["## Applying Mapper\n","\n","First step is to initialize a KeplerMapper object. You can ignore the `nerve` part."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"icpI6pbgH1Xe","executionInfo":{"status":"ok","timestamp":1636389898656,"user_tz":300,"elapsed":15,"user":{"displayName":"Sourabh Palande","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"18429378630788159865"}},"outputId":"c098bf08-28bf-4248-e1f7-2e552c53c267"},"source":["# Initialize mapper object\n","mymapper = km.KeplerMapper(verbose=1)\n","\n","# Define Nerve\n","nerve = km.GraphNerve(min_intersection=1)"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["KeplerMapper(verbose=1)\n"]}]},{"cell_type":"markdown","metadata":{"id":"shvxG8HrH1Xe"},"source":["### Define lens / filter function\n","\n","Next, we need to define the *lens*. In the python file `lenses.py`, I have defined a lens called `fsga_transform`. Given a factor and a specific level of that factor (for example, factor: stress, level: healthy), we construct a lens following the method described in __[Nicolau et. al. 2011](https://www.pnas.org/content/108/17/7265)__.\n","\n","For example, for factor: stress, and level: healthy, we take the gene expression profiles of all the *healthy* samples from the data and fit a linear model. Then we project all the samples on to this linear model and compute the residuals. The *lens* is the norm of the residual vector."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"aG6qmz4YH1Xe","executionInfo":{"status":"ok","timestamp":1636389907478,"user_tz":300,"elapsed":8714,"user":{"displayName":"Sourabh Palande","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"18429378630788159865"}},"outputId":"2b4d0ccf-b8a9-4e6a-dcbd-1f8dc5cfea24"},"source":["# Define lens\n","scaler = MinMaxScaler()\n","residuals, idx_tr, idx_te = fsga_transform(df, orthos, filter_by_factor, filter_by_level)\n","lens = mymapper.project(residuals, projection='l2norm', scaler=scaler)"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["..Projecting on data shaped (2671, 6335)\n","\n","..Projecting data using: l2norm\n","\n","..Scaling with: MinMaxScaler(copy=True, feature_range=(0, 1))\n","\n"]}]},{"cell_type":"markdown","metadata":{"id":"szrY0K5qH1Xf"},"source":["### Define cover\n","\n","Next step, define the cover. Just specify the number of intervals `cubes` and the amount of overlap between consecutive intervals `overlap` and let `kmapper` take care of it. Feel free to change both the parameters but keep in mind that overlap must be between 0 and 100. Also, increasing the number of intervals will make the algorithm run slower so don't increase it beyond 130 or so."]},{"cell_type":"code","metadata":{"id":"hu8AzP9hH1Xf"},"source":["# Define cover\n","cubes, overlap = (100, 85)\n","cover = km.cover.Cover(n_cubes=cubes, perc_overlap=overlap/100.)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"c2M_0muXH1Xf"},"source":["### Define clustering algorithm\n","\n","We will stick to DBSCAN with its default parameters.<br>\n","However, the metric we will use is the correlation distance (1 - correlation) between a pair of gene expression profiles. You can try changing this to *cosine* metric or some other predefined metric available in scikit-learn and see how it affects the output."]},{"cell_type":"code","metadata":{"id":"UDq-vX0XH1Xg"},"source":["# Define clustering algorithm\n","clust_metric = 'correlation'\n","clusterer = DBSCAN(metric=clust_metric)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"umU9bjJfH1Xg"},"source":["### Construct the mapper graph\n","\n","With all the components required to construct the mapper graph ready, we can go ahead and call the `map` method of the KepplerMapper object to construct the mapper graph. Keep an eye on the number of hypercubes, nodes and edges reported by the algorithm. You can change the graph size by changing the cover parameters."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"szoA2D8_H1Xg","executionInfo":{"status":"ok","timestamp":1636389921520,"user_tz":300,"elapsed":14054,"user":{"displayName":"Sourabh Palande","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"18429378630788159865"}},"outputId":"b84139cc-7362-46c7-b229-75a40cb5e98e"},"source":["# Create mapper 'graph' with nodes, edges and meta-information.\n","graph = mymapper.map(lens=lens,\n","                     X=residuals,\n","                     clusterer=clusterer,\n","                     cover=cover,\n","                     nerve=nerve,\n","                     precomputed=False,\n","                     remove_duplicate_nodes=True)"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Mapping on data shaped (2671, 6335) using lens shaped (2671, 1)\n","\n","Creating 100 hypercubes.\n","Merged 41 duplicate nodes.\n","\n","Number of nodes before merger: 361; after merger: 320\n","\n","\n","Created 1703 edges and 320 nodes in 0:00:14.192770.\n"]}]},{"cell_type":"markdown","metadata":{"id":"6tEZ0UsiH1Xh"},"source":["### Adding components to visualization\n","\n","Before we visualize the constructed mapper graph, we will add a couple of components to the visualization.<br>\n","First, we will color the nodes of the mapper graph using the specified factor (`color_by_factor`). The specified level (`color_by_level`) will be at one end of the colormap, all other levels will be at the other end. The node color is determined averaging the colors of all samples in the corresponding cluster."]},{"cell_type":"code","metadata":{"id":"1CkiBl_NH1Xh"},"source":["# Color nodes by specified color_by_factor, color_by_level\n","df[color_by_factor] = df[color_by_factor].astype('category')\n","color_vec = np.asarray([0 if(val == color_by_level) else 1 for val in df[color_by_factor]])\n","cscale = colorscale_from_matplotlib_cmap(plt.get_cmap('coolwarm'))"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"_80gw1jOH1Xh"},"source":["# show filter_by_factor levels in tooltip\n","temp = ['({}, {})'.format(str(p[0]), str(p[1])) for p in zip(df[filter_by_factor], df[color_by_factor])]\n","df['tooltips'] = temp"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Pr4T8JfCH1Xi"},"source":["### Visualize the mapper graph\n","\n","Latly, we create the visualization, save it as an html file, and then load it into a frame.<br>\n","Alternatively, you can browse to the html file and open it in a separate browser window."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":34},"id":"z6_qiRPRH1Xi","executionInfo":{"status":"ok","timestamp":1636389936573,"user_tz":300,"elapsed":15067,"user":{"displayName":"Sourabh Palande","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"18429378630788159865"}},"outputId":"822fcd9d-94a3-4f7f-ec3b-809df751d3ac"},"source":["# Specify file to save html output\n","fname = 'FilterBy_{}_ColorBy_{}_Cubes_{}_Overlap_{}.html'.format(filter_by_factor,\n","                                                              color_by_factor,\n","                                                              cubes,\n","                                                              overlap)\n","figtitle = 'Lens: {} : {}, Color by {} : {}, # Intervals {}, overlap {}'.format(filter_by_factor,\n","                                                                                filter_by_level,\n","                                                                                color_by_factor,\n","                                                                                color_by_level,\n","                                                                                cubes, overlap/100.0)\n","\n","fpath = projdir + '/' + fname\n","# Create visualization and save to specified file\n","_ = mymapper.visualize(graph,\n","                       path_html=fpath,\n","                       title=figtitle,\n","                       color_values=color_vec,\n","                       color_function_name=color_by_factor,\n","                       colorscale=cscale,\n","                       custom_tooltips=df['tooltips'])\n","\n","# Load the html output file\n","IFrame(src=fpath, width=1000, height=800)"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Wrote visualization to: /content/gdrive/MyDrive/PlantsAndPython-2021-10-22/FilterBy_family_ColorBy_tissue_Cubes_100_Overlap_85.html\n"]},{"output_type":"execute_result","data":{"text/html":["\n","        <iframe\n","            width=\"1000\"\n","            height=\"800\"\n","            src=\"/content/gdrive/MyDrive/PlantsAndPython-2021-10-22/FilterBy_family_ColorBy_tissue_Cubes_100_Overlap_85.html\"\n","            frameborder=\"0\"\n","            allowfullscreen\n","        ></iframe>\n","        "],"text/plain":["<IPython.lib.display.IFrame at 0x7f00fcfa2850>"]},"metadata":{},"execution_count":14}]}]}
